# Jobs
это объект, который управляет выполнением одноразовых задач, к примеру запуск скрипта для очистки базы данных, накатит миграции базы данных ( если это нужно сделать 1 или несколько раз, но не бесконечное кол-во, как в Deployment( т.к всего 1 политика перезапуска 'Always',к примеру ) после выполнения и Job и Pod НЕ удаляются
<details> <summary>0-job.yaml</summary>

``` 
apiVersion: batch/v1
kind: Job
metadata:
  name: pi
spec:
  template:
    spec:
      containers:
      - name: pi
        image: perl
        command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
```
</details>
Джоба написала 2000 знаков числа Пи и получила статус completed, в отличие от того же Deployment, который всегда running( пока не остановим )

![image](https://github.com/user-attachments/assets/95240909-adaf-44ec-ac6f-07d98d06937b)

Существуют 2 политики перезапуска Job:
- Never
- OnFailure

<details><summary>2-job-never.yaml</summary>

```
apiVersion: batch/v1
kind: Job
metadata:
  name: job-with-restartpolicy-never
spec:
  backoffLimit: 3              # кол-во повторных попыток, прежде чем Job будет считаться неудачной ( по умолчанию 6 )
  template:
    spec:
      containers:
      - name: job-with-failure
        image: busybox
        command: ["/bin/sh", "-c"]
        args: ["echo 'Running Job'; sleep 5; exit 1"]     # exit 1 это и есть ошибка для докера, т.к код отличный от 0
      restartPolicy: Never
```
</details>

Never контролируется job-controller, он не перезапускает ПОД(с ошибкой), а пересоздаёт новые ПОДЫ, и ПОДЫ после выполнения Job не удаляются(можно посмотреть причину ошибки)

![image](https://github.com/user-attachments/assets/51b097df-8eeb-46e8-8799-e56d778b3887)

<details><summary>1-job-onfailure.yaml</summary>

```
apiVersion: batch/v1
kind: Job
metadata:
  name: job-with-restartpolicy-onfailure
spec:
  backoffLimit: 3
  template:
    spec:
      containers:
      - name: job-with-failure
        image: busybox
        command: ["/bin/sh", "-c"]
        args: ["echo 'Running Job'; sleep 5; exit 1"]
      restartPolicy: OnFailure
```
</details>

OnFailure контролируется kubelet, он видит, что ПОД заканчивается с ошибкой и пытается перезапустить его, и после 3х попыток удаляет его(нельзя посмотреть причину ошибки), сама же Job остаётся, но со статусом failed. 

![image](https://github.com/user-attachments/assets/c50cc8f2-0208-4df2-8968-64ffc2451625)

